source/opt/merge_return_pass.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "merge_return_pass.h"
+
+#include "instruction.h"
+#include "ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status MergeReturnPass::Process(ir::IRContext *irContext) {
+  InitializeProcessing(irContext);
+
+  context()->InvalidateAnalysesExceptFor(ir::IRContext::kAnalysisDefUse);
+
+  bool modified = false;
+  for (auto &function : *get_module()) {
+    std::vector<ir::BasicBlock *> returnBlocks = collectReturnBlocks(&function);
+    modified |= mergeReturnBlocks(&function, returnBlocks);
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+std::vector<ir::BasicBlock *> MergeReturnPass::collectReturnBlocks(
+    ir::Function *function) {
+  std::vector<ir::BasicBlock *> returnBlocks;
+  for (auto &block : *function) {
+    ir::Instruction &terminator = *block.tail();
+    if (terminator.opcode() == SpvOpReturn ||
+        terminator.opcode() == SpvOpReturnValue) {
+      returnBlocks.push_back(&block);
+    }
+  }
+
+  return std::move(returnBlocks);
+}
+
+bool MergeReturnPass::mergeReturnBlocks(
```

Comment:
This is the right algorithm when structured control flow is not required.  If structured control flow is required, then you have to do a lot more work:  From within nested control flow, you have to branch out to the enclosing merge block; and you'll need to track whether you're on the "escape" path, via a bool flag.  It's complicated.... :-)